crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,584 @@
+                            );
+                        }
+                    }
+                }
+
+                Update::RemoveChunk(expected_chunk_identifier) => {
+                    let chunk_index = self
+                        .chunks
+                        .iter()
+                        .position(|(chunk_identifier, _)| {
+                            chunk_identifier == expected_chunk_identifier
+                        })
+                        // SAFETY: Assuming `LinkedChunk` and `ObservableUpdates` are not buggy, and
+                        // assuming `Self::chunks` is correctly initialized, it is not possible to
+                        // remove a chunk that does not exist. If this predicate fails, it means
+                        // `LinkedChunk` or `ObservableUpdates` contain a bug.
+                        .expect("Removing a chunk: The chunk is not found");
+
+                    // It's OK to ignore the result. The `chunk_index` exists because it's been
+                    // found, and we don't care about its associated value.
+                    let _ = self.chunks.remove(chunk_index);
+                }
+
+                Update::PushItems { at: position, items } => {
+                    let expected_chunk_identifier = position.chunk_identifier();
+
+                    let (chunk_index, offset, chunk_length) = {
+                        let control_flow = self.chunks.iter_mut().enumerate().try_fold(
+                            position.index(),
+                            |offset, (chunk_index, (chunk_identifier, chunk_length))| {
+                                if chunk_identifier == &expected_chunk_identifier {
+                                    ControlFlow::Break((chunk_index, offset, chunk_length))
+                                } else {
+                                    ControlFlow::Continue(offset + *chunk_length)
+                                }
+                            },
+                        );
+
+                        match control_flow {
+                            // Chunk has been found, and all values have been calculated as
+                            // expected.
+                            ControlFlow::Break(values) => values,
+
+                            // Chunk has not been found.
+                            ControlFlow::Continue(..) => {
+                                // SAFETY: Assuming `LinkedChunk` and `ObservableUpdates` are not
+                                // buggy, and assuming `Self::chunks` is correctly initialized, it
+                                // is not possible to push items on a chunk that does not exist. If
+                                // this predicate fails, it means `LinkedChunk` or
+                                // `ObservableUpdates` contain a bug.
+                                panic!("Pushing items: The chunk is not found");
+                            }
+                        }
```

Comment:
I still find this a bit obtuse and harder to read for newcomers to this code, would really prefer a plain loop with a `let mut found = false` flag :pensive: